File: mcp-server/src/tools/usce-check.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Atomira Technologies, S.L.

import { ALL_CARDS, findPrincipleCard } from '../cards/seed.js';
import { runUsceChecks } from '../cards/usce.js';
import { renderVerdict } from './render.js';
import type { McpTool } from './types.js';

export const usceCheckTool: McpTool = {
  name: 'usce_check',
  description:
    'Run the Universal Sanity Check Engine (USCE) on a finished numeric output: range-check the values against a principle card\'s validationEnvelopes. Pass `id` (a principle card id, e.g. "ideal-gas-law") and `output` (a map of envelope key to number, e.g. {"gasConstant_J_per_molK": 8.3145}). Returns per-key verdicts (within -> pass, outside -> HIGH) and an overall severity. v1 checks validation envelopes; causality and asymptotic-decay checks over time-series outputs are future work.',
  inputSchema: {
    type: 'object',
    properties: {
      id: {
        type: 'string',
        description:
          'Principle card id whose validationEnvelopes to check against, e.g. "ideal-gas-law".',
      },
      output: {
        type: 'object',
        additionalProperties: { type: 'number' },
        description:
          'Map of envelope key to the finished output value, e.g. { "gasConstant_J_per_molK": 8.3145 }.',
      },
    },
    required: ['id', 'output'],
  },
  async run(input) {
    const id = String(input.id ?? '').trim();
    if (!id) {
      throw new Error('Empty id.');
    }
    const card = findPrincipleCard(id);
    if (!card) {
      const known = ALL_CARDS.map((c) => c.id).join(', ');
      throw new Error(
        `No principle card with id "${id}" in the seed corpus. Known ids: ${known}.`,
      );
    }
    const output = (input.output ?? {}) as Record<string, number>;
    // Shared renderer: four tools now format a verdict, and four copies of the
    // same layout is four places for the wording to drift invisibly, since each
    // tool is tested against its own output.
    return renderVerdict('USCE verdict', card, runUsceChecks(output, card));
  },
};

File: examples/usce_check.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 Atomira Technologies, S.L.

"""
USCE — check a finished output against a card's validation envelopes.

The cross-check engine vets a *proposed* formula. USCE vets a *finished
result*: it range-checks the numbers your code produced against the [min, max]
bounds the card declares (within -> pass, outside -> HIGH). No database, no
API keys.

Run:
    pip install -e ../sdk-py
    python usce_check.py
"""
from artano_lemma import find_card, load_cards, run_usce_checks

corpus = load_cards()
card = find_card("ideal-gas-law", corpus)  # declares gasConstant_J_per_molK in [8.314, 8.315]


def report(label: str, output: dict) -> None:
    result = run_usce_checks(output, card)
    o = result.overall
    print(f"\n{label}: {o.severity}  ({o.passing}/{o.total} checks pass)")
    for check in result.checks:
        print(f"  [{check.severity}] {check.detail}")
    print(f"  -> {result.diagnosis}")


# A correct result: the gas constant the code used is in range.
report("R = 8.3145 (correct)", {"gasConstant_J_per_molK": 8.3145})

# A wrong result: the code used a bad constant -> USCE catches it.
report("R = 9.0 (wrong)", {"gasConstant_J_per_molK": 9.0})
